Jetty 12 inserted handler in ee10 servlet context

gregw opened the pull request:
This PR refactors the ee10 handing of servlet API request and response objects:
- The `ServletContextHandler` matches the request to a servlet and creates a one time only `ServletContextRequest` and a `ServletContextResponse`
- A reusable `ServletChannel` object with all the heavy weight `HttpInput` and `HttpOutput` object is associated with the `ServletContextRequest` and `ServletContextResponse`.
- Once the handling reaches the `ServletHandler`, the possibly wrapped request, response and callback are associated with the `ServletChannel` before handling.
- Were possible the `ServletApiRequest` and `ServletApiResponse` use the possibly wrapped request/response

I have added tests to check that GzipHandler can now be nested inside of an EE10 context.

gregw committed:
Added distro test for static content features.

lorban committed:
fix ee9 DefaultServlet to make it able to send static content that isn't backed by a Path

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

gregw committed:
Fixed favicon.ico test

gregw committed:
Disabled gzipHandler for ee10 demo

gregw committed:
Wrap context with gzipHandler for ee10 test

gregw committed:
Misc cleanups of servlet request/response wrappers to simplify them.

gregw committed:
Merge branch 'jetty-12.0.x' into jetty-12-ee10-late-servlet-wrap

gregw committed:
WIP

Signed-off-by: gregw <gregw@webtide.com>

gregw committed:
ServletChannel holds HttpOutput

gregw committed:
WIP

gregw committed:
WIP

gregw committed:
WIP

gregw committed:
WIP

gregw committed:
WIP

gregw committed:
WIP

gregw committed:
Testing gzip in context

gregw committed:
Revert "Disabled gzipHandler for ee10 demo"

This reverts part of commit 12b7e7b0

lorban commented on jetty-core/jetty-websocket/jetty-websocket-core-server/src/main/java/org/eclipse/jetty/websocket/core/server/internal/WebSocketHttpFieldsWrapper.java:
```diff
@@ -14,75 +14,120 @@
+    public HttpField onAddField(HttpField field)
     {
-        if (HttpHeader.SEC_WEBSOCKET_SUBPROTOCOL.is(name))
+        System.err.println("add " + field);
```
Change that `System.err.println` to a `Log.debug`

janbartel commented:
@gregw can you edit the initial post to include a sentence or two on the motivation for the PR?

lorban commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletContextRequest.java:
```diff
@@ -269,22 +264,19 @@ ServletChannel getServletChannel()
     public ServletApiRequest getServletApiRequest()
     {
+        if (_servletChannel.getCallback() == null)
+            new RuntimeException("getServletApiRequest called before ServletHandler").printStackTrace();
```
Do you want to keep some form of debug logging here or should that become an assertion throwing an exception?

lorban commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletContextResponse.java:
```diff
@@ -170,60 +162,65 @@ protected ServletApiResponse newServletApiResponse()
     }
 
-    public boolean isContentComplete(long written)
+    public boolean isContentInComplete(long written)
```
Since you mean 'content not yet complete', you shouldn't capitalize `Complete`.
I.e.: this should be renamed `isContentIncomplete`.

lorban commented on jetty-ee10/jetty-ee10-servlet/src/test/java/org/eclipse/jetty/ee10/servlet/AsyncServletTest.java:
```diff
@@ -99,7 +100,20 @@ public void setUp() throws Exception
-        _server.setHandler(context);
+
+        _latch = new CountDownLatch(1);
+        EventsHandler events = new EventsHandler()
```
You could pass the wrapped handler in `EventsHandler`'s constructor instead of calling `setHandler`:
```java
EventsHandler events = new EventsHandler(context)
{
 ...
}
_server.setHandler(events);
```

gregw replied:
Hmmmm sometimes we pass in the parent in the constructor.  If we starting passing the child it could get confusing.

janbartel commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletHandler.java:
```diff
@@ -453,9 +453,9 @@ protected IdentityService getIdentityService()
     public boolean handle(Request request, Response response, Callback callback) throws Exception
     {
         // We will always have a ServletScopedRequest and MappedServlet otherwise we will not reach ServletHandler.
```
What's a `ServletScopedRequest`?

gregw replied:
improved the comments

janbartel reviewed (changes requested):
This is a trivial javadoc error. However, it should be fixed. Also, for a more thorough review, I think we need a hangout.

lorban commented on jetty-ee10/jetty-ee10-servlet/src/test/java/org/eclipse/jetty/ee10/servlet/AsyncServletLongPollTest.java:
```diff
@@ -130,6 +130,7 @@ protected void doDelete(HttpServletRequest request, HttpServletResponse response
 
                 HttpTester.Input input2 = HttpTester.from(socket2.getInputStream());
                 HttpTester.Response response2 = HttpTester.parseResponse(input2);
+                System.err.println(response2);
```
To be removed.

lorban commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletContextResponse.java:
```diff
@@ -170,60 +162,65 @@ protected ServletApiResponse newServletApiResponse()
     {
-        return _httpOutput;
+        if (_servletChannel.getCallback() == null)
+            new RuntimeException("getHttpOutput called before ServletHandler").printStackTrace();
```
Should that be kept as a debug log or be changed to an assertion throwing an exception?

lorban commented on jetty-ee10/jetty-ee10-tests/jetty-ee10-test-integration/src/test/resources/add-jetty-test-webapp.xml:
```diff
@@ -3,16 +3,18 @@
+        <New class="org.eclipse.jetty.ee10.webapp.WebAppContext">
+          <Set name="contextPath">/test-jetty-webapp</Set>
+          <Set name="war"><Property name="test.webapps" default="." />/test-jetty-webapp.war</Set>
+          <Call name="insertHandler">
```
What's the point of this `insertHandler` call?

gregw replied:
old

gregw committed:
improved comments

lorban reviewed (commented):
1st review pass

gregw committed:
updates from review

gregw committed:
fixed javadoc

lachlan-roberts commented on jetty-core/jetty-websocket/jetty-websocket-core-server/src/main/java/org/eclipse/jetty/websocket/core/server/internal/WebSocketHttpFieldsWrapper.java:
```diff
@@ -14,75 +14,118 @@
+
+                case SEC_WEBSOCKET_EXTENSIONS ->
+                {
+                    // TODO: why add extensions??
```
What do you mean by this TODO?

gregw replied:
No idea, it was there before. I'll remove.

lachlan-roberts commented on jetty-core/jetty-websocket/jetty-websocket-core-server/src/main/java/org/eclipse/jetty/websocket/core/server/internal/WebSocketHttpFieldsWrapper.java:
```diff
@@ -14,75 +14,118 @@
+                case SEC_WEBSOCKET_EXTENSIONS ->
+                {
+                    // TODO: why add extensions??
+                    _response.setExtensions(Collections.emptyList());
```
This is not strictly correct as it is removing 1 `HttpField` at a time and we could have multiple `SEC_WEBSOCKET_EXTENSIONS` fields.

Maybe we need a `_response.removeExtension(...)` for this.

gregw replied:
true.    Note that it is mostly not used due to the wrapping of the put methods.  But if an app was to try to do a specific remove....
I have put in a TODO for now, as the problem existed before.   For another PR

lachlan-roberts replied:
I don't think this problem did exist before, there was this method which only took the header name, and it was called before removing all `HttpFields` for a `HttpHeader`, not for removing an individual `HttpField`.
```java
public boolean onRemoveField(String name)
```

gregw replied:
@lachlan-roberts ah the problem before was that the HttpFields wrapper did not handle at all removal of an individual field as it did not modify the iterator.  So it just ignored it.    I'll see if I can fix then....

gregw replied:
@lachlan-roberts how about this now?

lachlan-roberts replied:
Yep that looks like it will work now.

lachlan-roberts commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpFields.java:
```diff
@@ -1204,6 +1204,167 @@ else if (ensured > 0)
+             * @param field The field being added.
+             * @return The field to add, or null if the add is to be ignored.
+             */
+            public HttpField onAddField(HttpField field)
```
The `onAddField()` method is being called twice for each add.
Once in `Wrapper.add()` and the other in `ListIterator` implementation.

lachlan-roberts commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpFields.java:
```diff
@@ -1204,6 +1204,167 @@ else if (ensured > 0)
+        }
     }
 
     /**
```
Also you should fix the `hashCode()` method on `MutableHttpFields`. 
It is iterating over the array length and not `_size` so will throw NPE if called.

gregw committed:
updates from review

gregw committed:
update javadoc

gregw committed:
fixed websocket remove extension

gregw committed:
Added info interfaces

gregw commented:
@janbartel @lorban @lachlan-roberts I've now added `ServletRequestInfo` and `ServletResponseInfo` interfaces, which are how the ServletApiRequest/ServletApiResponse communicate with the ServletContextRequest/ServletResponse.

gregw committed:
Added info interfaces

lorban commented on jetty-ee10/jetty-ee10-servlet/src/test/java/org/eclipse/jetty/ee10/servlet/GzipHandlerTest.java:
```diff
@@ -315,13 +324,26 @@ protected void doGet(HttpServletRequest request, HttpServletResponse response) t
+    }
+
+    @ParameterizedTest
+    @MethodSource("tests")
```
That can be replaced with:
```java
@ValueSource(booleans = {true, false})
```

lorban commented on jetty-ee10/jetty-ee10-servlet/src/test/java/org/eclipse/jetty/ee10/servlet/GzipHandlerTest.java:
```diff
@@ -372,9 +396,11 @@ public void testBlockingResponse() throws Exception
-    @Test
-    public void testGzipNotModifiedVaryHeader() throws Exception
+    @ParameterizedTest
+    @MethodSource("tests")
```
That can be replaced with:
```java
@ValueSource(booleans = {true, false})
```

lorban commented on jetty-ee10/jetty-ee10-servlet/src/test/java/org/eclipse/jetty/ee10/servlet/GzipHandlerTest.java:
```diff
@@ -410,9 +436,11 @@ public void testGzipNotModifiedVaryHeader() throws Exception
-    @Test
-    public void testAsyncResponse() throws Exception
+    @ParameterizedTest
+    @MethodSource("tests")
```
That can be replaced with:
```java
@ValueSource(booleans = {true, false})
```

lorban commented on jetty-ee10/jetty-ee10-servlet/src/test/java/org/eclipse/jetty/ee10/servlet/GzipHandlerTest.java:
```diff
@@ -436,9 +464,11 @@ public void testAsyncResponse() throws Exception
-    @Test
-    public void testBufferResponse() throws Exception
+    @ParameterizedTest
+    @MethodSource("tests")
```
That can be replaced with:
```java
@ValueSource(booleans = {true, false})
```

lorban commented on jetty-ee10/jetty-ee10-servlet/src/test/java/org/eclipse/jetty/ee10/servlet/GzipHandlerTest.java:
```diff
@@ -462,9 +492,11 @@ public void testBufferResponse() throws Exception
-    @Test
-    public void testAsyncLargeResponse() throws Exception
+    @ParameterizedTest
+    @MethodSource("tests")
```
That can be replaced with:
```java
@ValueSource(booleans = {true, false})
```

lorban commented on jetty-ee10/jetty-ee10-servlet/src/test/java/org/eclipse/jetty/ee10/servlet/GzipHandlerTest.java:
```diff
@@ -494,9 +526,11 @@ public void testAsyncLargeResponse() throws Exception
-    @Test
-    public void testAsyncEmptyResponse() throws Exception
+    @ParameterizedTest
+    @MethodSource("tests")
```
That can be replaced with:
```java
@ValueSource(booleans = {true, false})
```

lorban commented on jetty-ee10/jetty-ee10-servlet/src/test/java/org/eclipse/jetty/ee10/servlet/GzipHandlerTest.java:
```diff
@@ -517,9 +551,11 @@ public void testAsyncEmptyResponse() throws Exception
-    @Test
-    public void testGzipHandlerWithMultipleAcceptEncodingHeaders() throws Exception
+    @ParameterizedTest
+    @MethodSource("tests")
```
That can be replaced with:
```java
@ValueSource(booleans = {true, false})
```

lorban commented on jetty-ee10/jetty-ee10-servlet/src/test/java/org/eclipse/jetty/ee10/servlet/GzipHandlerTest.java:
```diff
@@ -545,9 +581,11 @@ public void testGzipHandlerWithMultipleAcceptEncodingHeaders() throws Exception
-    @Test
-    public void testGzipNotMicro() throws Exception
+    @ParameterizedTest
+    @MethodSource("tests")
```
That can be replaced with:
```java
@ValueSource(booleans = {true, false})
```

lorban commented on jetty-ee10/jetty-ee10-servlet/src/test/java/org/eclipse/jetty/ee10/servlet/GzipHandlerTest.java:
```diff
@@ -572,9 +610,11 @@ public void testGzipNotMicro() throws Exception
-    @Test
-    public void testGzipNotMicroChunked() throws Exception
+    @ParameterizedTest
+    @MethodSource("tests")
```
That can be replaced with:
```java
@ValueSource(booleans = {true, false})
```

lorban commented on jetty-ee10/jetty-ee10-servlet/src/test/java/org/eclipse/jetty/ee10/servlet/GzipHandlerTest.java:
```diff
@@ -601,9 +641,11 @@ public void testGzipNotMicroChunked() throws Exception
-    @Test
-    public void testETagNotGzipHandler() throws Exception
+    @ParameterizedTest
+    @MethodSource("tests")
```
That can be replaced with:
```java
@ValueSource(booleans = {true, false})
```

lorban commented on jetty-ee10/jetty-ee10-servlet/src/test/java/org/eclipse/jetty/ee10/servlet/GzipHandlerTest.java:
```diff
@@ -622,9 +664,11 @@ public void testETagNotGzipHandler() throws Exception
-    @Test
-    public void testETagGzipHandler() throws Exception
+    @ParameterizedTest
+    @MethodSource("tests")
```
That can be replaced with:
```java
@ValueSource(booleans = {true, false})
```

lorban commented on jetty-ee10/jetty-ee10-servlet/src/test/java/org/eclipse/jetty/ee10/servlet/GzipHandlerTest.java:
```diff
@@ -643,9 +687,11 @@ public void testETagGzipHandler() throws Exception
-    @Test
-    public void testDeleteETagGzipHandler() throws Exception
+    @ParameterizedTest
+    @MethodSource("tests")
```
That can be replaced with:
```java
@ValueSource(booleans = {true, false})
```

lorban commented on jetty-ee10/jetty-ee10-servlet/src/test/java/org/eclipse/jetty/ee10/servlet/GzipHandlerTest.java:
```diff
@@ -675,9 +721,11 @@ public void testDeleteETagGzipHandler() throws Exception
-    @Test
-    public void testForwardGzipHandler() throws Exception
+    @ParameterizedTest
+    @MethodSource("tests")
```
That can be replaced with:
```java
@ValueSource(booleans = {true, false})
```

lorban commented on jetty-ee10/jetty-ee10-servlet/src/test/java/org/eclipse/jetty/ee10/servlet/GzipHandlerTest.java:
```diff
@@ -702,9 +750,11 @@ public void testForwardGzipHandler() throws Exception
-    @Test
-    public void testIncludeGzipHandler() throws Exception
+    @ParameterizedTest
+    @MethodSource("tests")
```
That can be replaced with:
```java
@ValueSource(booleans = {true, false})
```

lorban commented on jetty-ee10/jetty-ee10-servlet/src/test/java/org/eclipse/jetty/ee10/servlet/GzipHandlerTest.java:
```diff
@@ -729,11 +779,14 @@ public void testIncludeGzipHandler() throws Exception
-    @Test
-    public void testIncludeExcludeGzipHandlerInflate() throws Exception
+    @ParameterizedTest
+    @MethodSource("tests")
```
That can be replaced with:
```java
@ValueSource(booleans = {true, false})
```

lorban commented on jetty-ee10/jetty-ee10-servlet/src/test/java/org/eclipse/jetty/ee10/servlet/GzipHandlerTest.java:
```diff
@@ -784,9 +837,11 @@ public void testAddGetPaths()
-    @Test
-    public void testGzipRequest() throws Exception
+    @ParameterizedTest
+    @MethodSource("tests")
```
That can be replaced with:
```java
@ValueSource(booleans = {true, false})
```

lorban commented on jetty-ee10/jetty-ee10-servlet/src/test/java/org/eclipse/jetty/ee10/servlet/GzipHandlerTest.java:
```diff
@@ -816,9 +871,11 @@ public void testGzipRequest() throws Exception
-    @Test
-    public void testGzipRequestChunked() throws Exception
+    @ParameterizedTest
+    @MethodSource("tests")
```
That can be replaced with:
```java
@ValueSource(booleans = {true, false})
```

lorban commented on jetty-ee10/jetty-ee10-servlet/src/test/java/org/eclipse/jetty/ee10/servlet/GzipHandlerTest.java:
```diff
@@ -848,9 +905,11 @@ public void testGzipRequestChunked() throws Exception
-    @Test
-    public void testGzipFormRequest() throws Exception
+    @ParameterizedTest
+    @MethodSource("tests")
```
That can be replaced with:
```java
@ValueSource(booleans = {true, false})
```

lorban commented on jetty-ee10/jetty-ee10-servlet/src/test/java/org/eclipse/jetty/ee10/servlet/GzipHandlerTest.java:
```diff
@@ -876,9 +935,11 @@ public void testGzipFormRequest() throws Exception
-    @Test
-    public void testGzipBomb() throws Exception
+    @ParameterizedTest
+    @MethodSource("tests")
```
That can be replaced with:
```java
@ValueSource(booleans = {true, false})
```

lorban commented on jetty-ee10/jetty-ee10-servlet/src/test/java/org/eclipse/jetty/ee10/servlet/GzipHandlerTest.java:
```diff
@@ -907,9 +968,11 @@ public void testGzipBomb() throws Exception
-    @Test
-    public void testGzipExcludeNewMimeType() throws Exception
+    @ParameterizedTest
+    @MethodSource("tests")
```
That can be replaced with:
```java
@ValueSource(booleans = {true, false})
```

lorban commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletRequestState.java:
```diff
@@ -311,7 +311,7 @@ public boolean abortResponse()
 
                 case OPEN:
-                    _servletChannel.getResponse().setStatus(500);
+                    _servletChannel.getServletContextResponse().setStatus(500);
```
Replace the `500` with `HttpStatus.INTERNAL_SERVER_ERROR_500`

lorban reviewed (dismissed):
LGTM besides a few nits.

gregw committed:
Merge branch 'jetty-12.0.x' into jetty-12-ee10-late-servlet-wrap

# Conflicts:
#	jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletChannel.java
#	jetty-ee10/jetty-ee10-servlet/src/test/java/org/eclipse/jetty/ee10/servlet/AsyncServletTest.java

gregw committed:
updates from review

gregw committed:
fixed merge

gregw committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into jetty-12-ee10-late-servlet-wrap

gregw commented:
@janbartel @lorban @lachlan-roberts I think the test failure is a flake.   So can I get a review on this before tonight's meeting?

gregw committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into jetty-12-ee10-late-servlet-wrap

janbartel commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletContextRequest.java:
```diff
@@ -58,7 +65,7 @@ public class ServletContextRequest extends ContextRequest
     {
         if (request instanceof ServletApiRequest)
-            return ((ServletApiRequest)request).getServletContextRequest();
+            return ((ServletApiRequest)request).getServletRequestInfo().getServletContextRequest();
```
This seems odd. The ServletRequestInfo interface is designed so we don't have to reference the ServletContextRequest directly, and is implemented by the ServletContextRequest. So seems odd that the interface has this method. Is it really necessary?

gregw replied:
We do need a way to navigate back from a `ServletApiRequest` to the `ServletContextRequest`,  But I have now made that only be `ServletContextRequest.getServletContextRequest(ServletRequest)`

janbartel reviewed (dismissed):
A bit of an oddity with the `ServletRequestInfo.getServletContextRequest()` method, but other than that, +1.

gregw committed:
Merge branch 'jetty-12.0.x' into jetty-12-ee10-late-servlet-wrap

gregw committed:
updates from review

removed explicit method to navigate from Request/Response Info back to core context request/response

gregw commented:
@janbartel @lachlan-roberts @lorban sorry, but after that last little tweak, I need a review from at least 1 of you again

grgrzybek commented:
@gregw I can't track the exact part of 0b1c28a8883bb70421a228d1a72644d6125f2baa, but I think it started causing NPE in some special case.

My servlet is invoked with `/gateway/x?what=include&where=/`:
```java
String what = req.getParameter("what");
String where = req.getParameter("where");
switch (what) {
	case "redirect":
		resp.sendRedirect(where);
		return;
	case "forward":
		// we can't send anything when forwarding
		req.getRequestDispatcher(where).forward(req, resp);
		return;
	case "include":
		resp.getWriter().print(">>>");
		req.getRequestDispatcher(where).include(req, resp);
		resp.getWriter().print("<<<");
		return;
	default:
}
```

`org.eclipse.jetty.ee10.servlet.ServletApiResponse#getWriter()` calls:
```java
String encoding = getServletResponseInfo().getCharacterEncoding(true);
```

`org.eclipse.jetty.ee10.servlet.ServletContextResponse#getCharacterEncoding(boolean)` calls:
```java
if (setContentType)
    setCharacterEncoding(encoding, EncodingFrom.DEFAULT);
```

so `org.eclipse.jetty.ee10.servlet.ServletContextResponse#_characterEncoding` is set to `iso-8859-1`.
Because the forward is for `/`, default servlet is involved (because I need to test welcome files).

`org.eclipse.jetty.ee10.servlet.DefaultServlet#doGet()` has this fragment:
```java
if (contextResponse != null)
{
    String characterEncoding = contextResponse.getRawCharacterEncoding();
    if (characterEncoding != null)
        content = new ForcedCharacterEncodingHttpContent(content, characterEncoding);
}
```

and the constructor is:
```java
public ForcedCharacterEncodingHttpContent(HttpContent content, String characterEncoding)
{
    super(content);
    this.characterEncoding = characterEncoding;
    String mimeType = content.getContentTypeValue();
    int idx = mimeType.indexOf(";charset");
    if (idx >= 0)
        mimeType = mimeType.substring(0, idx);
    this.contentType = mimeType + ";charset=" + this.characterEncoding;
}
```

the `content` is:
```
content = {org.ops4j.pax.web.service.jetty.internal.web.DefaultServlet$UnknownLengthHttpContent@4857} "UnknownLengthHttpContent@c898d8f[CompressedFormatsHttpContent@1c61db46[ResourceHttpContent@471a9af8{r=file:/data/sources/github.com/ops4j/org.ops4j.pax.web-jakarta-new/pax-web-jetty/target/b1/,ct=null}]]"
 _delegate: org.eclipse.jetty.http.content.HttpContent  = {org.eclipse.jetty.http.content.PreCompressedHttpContentFactory$CompressedFormatsHttpContent@4919} "CompressedFormatsHttpContent@1c61db46[ResourceHttpContent@471a9af8{r=file:/data/sources/github.com/ops4j/org.ops4j.pax.web-jakarta-new/pax-web-jetty/target/b1/,ct=null}]"
  _delegate: org.eclipse.jetty.http.content.HttpContent  = {org.eclipse.jetty.http.content.ResourceHttpContent@4922} "ResourceHttpContent@471a9af8{r=file:/data/sources/github.com/ops4j/org.ops4j.pax.web-jakarta-new/pax-web-jetty/target/b1/,ct=null}"
   _contentType: java.lang.String  = null
   _etag: org.eclipse.jetty.http.HttpField  = {org.eclipse.jetty.http.PreEncodedHttpField@4925} "ETag: W/"///+d0tLiXs/////73HTsQ""
   _path: java.nio.file.Path  = {sun.nio.fs.UnixPath@4924} "/data/sources/github.com/ops4j/org.ops4j.pax.web-jakarta-new/pax-web-jetty/target/b1"
   _resource: org.eclipse.jetty.util.resource.Resource  = {org.eclipse.jetty.util.resource.PathResource@4749} "file:/data/sources/github.com/ops4j/org.ops4j.pax.web-jakarta-new/pax-web-jetty/target/b1/"
  compressedFormats: java.util.Set  = {java.util.HashSet@4921}  size = 0
```

the underlying resource is directory, so the extension is `*` and the resulting content type is null.

so I'm getting NullPointerException.

My test you can run (without bothering with OSGi stuff) is https://github.com/ops4j/org.ops4j.pax.web/blob/1b26a2f8ea394d696a6780ff0fa92b2613019dff/pax-web-jetty/src/test/java/org/ops4j/pax/web/service/jetty/internal/UnifiedJettyTest.java#L361
